### ISSUE TITLE
Incorrect constraint warning for nullable type argument

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#29067]
The following results in a warning for `A<B?>`.
```c#
[module: System.Runtime.CompilerServices.NonNullTypes(true)]

interface I<T> { }

class A<T> where T : I<T> { }

class B : I<B> { }

class Program
{
    static void Main()
    {
        new A<B>();
        new A<B?>(); // incorrect warning
    }
}
```
```
(14,15): warning CS8631: The type 'B?' cannot be used as type parameter 'T' in the generic type
    or method 'A<T>'. Nullability of type argument 'B?' doesn't match constraint type 'I<B?>'.
```

**Comment by jcouv** [comment]
@cston Looking at this issue again, I think the behavior is expected. 
`where T : I<T>` means `T` cannot be null, so it cannot be substituted with `B?`.
Am I missing something?
